modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/ObservationToRecordMapperImpl.kt

```diff
@@ -0,0 +1,83 @@
+package edu.stanford.healthconnectonfhir
+
+import androidx.health.connect.client.records.BloodPressureRecord
+import androidx.health.connect.client.records.HeartRateRecord
+import androidx.health.connect.client.records.Record
+import androidx.health.connect.client.records.WeightRecord
+import androidx.health.connect.client.units.Mass
+import androidx.health.connect.client.units.Pressure
+import org.hl7.fhir.r4.model.Observation
+import java.time.ZoneOffset
+import javax.inject.Inject
+
+class ObservationToRecordMapperImpl @Inject constructor() : ObservationToRecordMapper {
+
+    private val codeToMappingFunction: Map<String, (Observation) -> Record> = mapOf(
+        Loinc.BloodPressure.CODE to { obs -> mapToBloodPressureRecord(obs) },
+        Loinc.Weight.CODE to { obs -> mapToWeightRecord(obs) },
+        Loinc.HeartRate.CODE to { obs -> mapToHeartRateRecord(obs) }
+    )
+
+    override fun map(observation: Observation): Record {
```

Comment:
I would introduce a generic type here:

```kotlin
   override fun <T : Record> map(observation: Observation): T {
        @Suppress("UNCHECKED_CAST")
        return when(observation.code.codingFirstRep.code) {
            Loinc.BloodPressure.CODE -> mapToBloodPressureRecord(observation)
            Loinc.Weight.CODE -> mapToWeightRecord(observation)
            Loinc.HeartRate.CODE -> mapToHeartRateRecord(observation)
            else -> error("")
        } as T
    }
```

and reflect the same generic type in the observation repository as well

Reply:
Good idea. I like this approach. 💯